File: src/lib/memoryLog.ts
import { invoke } from "@tauri-apps/api/core";

export type MemoryLogRow = {
  id: number;
  scope: "project" | "global" | string;
  date: string;
  text: string;
  storePath: string;
};

export type MemoryLogPage = {
  rows: MemoryLogRow[];
  total: number;
  projectStorePath: string | null;
  globalStorePath: string;
};

export type MemoryLogFilter = {
  scope?: string;
  search?: string;
  repoRoot?: string;
  limit?: number;
  offset?: number;
};

export function queryMemoryLog(filter: MemoryLogFilter): Promise<MemoryLogPage> {
  return invoke<MemoryLogPage>("memory_log_query", { filter });
}

File: src-tauri/src/domain/memory_log.rs
//! Types for the read-only memory viewer — raw OptMem log rows from
//! project and global stores.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryLogRow {
    pub id: u32,
    pub scope: String,
    pub date: String,
    pub text: String,
    pub store_path: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryLogPage {
    pub rows: Vec<MemoryLogRow>,
    pub total: u32,
    pub project_store_path: Option<String>,
    pub global_store_path: String,
}

/// Filter for the memory viewer. `scope` is `"project"`, `"global"`, or
/// omitted for both. `search` is a case-insensitive substring match on the
/// raw note text.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryLogFilter {
    #[serde(default)]
    pub scope: Option<String>,
    #[serde(default)]
    pub search: Option<String>,
    #[serde(default)]
    pub repo_root: Option<String>,
    #[serde(default)]
    pub limit: Option<u32>,
    #[serde(default)]
    pub offset: Option<u32>,
}
